Add RequiresAssemblyFiles attribute

agocke opened the pull request:
This PR annotates every API doc'd as having different
behavior in single-file publishing with RequiresAssemblyFiles.
Some APIs might be special-cased by the analyzer to produce
special messages, but RAF is useful for visual inspection
of the APIs. Module.Name and .FullyQualifiedName were also
unannotated.

Docs: https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file

Assembly.Location | Returns an empty string.
Module.FullyQualifiedName | Returns a string with the value of `<Unknown>` or throws an exception.
Module.Name | Returns a string with the value of `<Unknown>`.
Assembly.GetFile | Throws IOException.
Assembly.GetFiles | Throws IOException.
Assembly.CodeBase | Throws PlatformNotSupportedException.
Assembly.EscapedCodeBase | Throws PlatformNotSupportedException.
AssemblyName.CodeBase | Returns null.
AssemblyName.EscapedCodeBase | Returns null.

agocke committed:
Add RequiresAssemblyFiles attribute

This PR annotates every API doc'd as having different
behavior in single-file publishing with RequiresAssemblyFiles.
Some APIs might be special-cased by the analyzer to produce
special messages, but RAF is useful for visual inspection
of the APIs. Module.Name and .FullyQualifiedName were also
unannotated.

agocke committed:
Add some more attributes in Mono

agocke committed:
Don't compile for netfx

jkotas commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -11058,15 +11058,17 @@ public sealed partial class AmbiguousMatchException : System.SystemException
     }
     public abstract partial class Assembly : System.Reflection.ICustomAttributeProvider, System.Runtime.Serialization.ISerializable
     {
+        internal const string ThrowingMessageInRAF = "This member throws an exception for assemblies embedded in a single-file app";
```
This won't be preserved next time somebody tries to auto-generate System.Runtime https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/updating-ref-source.md . 

These ref files should use inline strings since it is what the auto-generator produces.

jkotas commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Assemblies/RoAssembly.Modules.cs:
```diff
@@ -16,6 +17,9 @@ internal abstract partial class RoAssembly
         public sealed override Module? GetModule(string name) => GetRoModule(name);
         public sealed override Module[] GetModules(bool getResourceModules) => ComputeRoModules(getResourceModules).CloneArray<Module>();
 
+#if NETCOREAPP
```
I think this should be `#if NET5_0_OR_GREATER`

agocke committed:
Adjust ifdef

agocke committed:
Adjust generated code

jeffhandley commented:
@buyaa-n This PR is assigned to you for follow-up before the RC1 snap.

agocke committed:
Missed a ref file

agocke commented:
@buyaa-n anything else?

buyaa-n commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -11060,13 +11060,13 @@ public abstract partial class Assembly : System.Reflection.ICustomAttributeProvi
         protected Assembly() { }
         [System.ObsoleteAttribute("Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead.", DiagnosticId = "SYSLIB0012", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
-        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFilesAttribute("The code will throw for assemblies embedded in a single-file app")]
+        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFiles("This member throws an exception for assemblies embedded in a single-file app")]
```
Here and in all ref updates shouldn't we put the complete name for the attribute? At least all other attributes are written in full

buyaa-n commented on src/mono/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -410,6 +410,9 @@ public override Assembly GetSatelliteAssembly(CultureInfo culture, Version? vers
 
+        internal const string UnknownStringMessageInRAF = "Returns <Unknown> for modules with no file path";
+
+        [RequiresAssemblyFiles(UnknownStringMessageInRAF)]
```
The message is different than ref `"This member throws an exception for assemblies embedded in a single-file app"`

agocke replied:
You're right, these two have the wrong message.

buyaa-n reviewed (approved):
Left two minor comments, overall LGTM

agocke committed:
Address PR comments and add more annotation to the ref

agocke committed:
Found more missing annotations, clean up message and casing

agocke commented:
@buyaa-n If you wouldn't mind taking another look, I found some more missing annotations, and an inconsistency in casing that I've tried to fix up.

@VSadov @vitek-karas for change in casing in runtime. In theory this could be a breaking change for people upgrading from .NET 5 to .NET 6 but I think the inconsistent casing meant it was impossible for them to rely on specific behavior in this situation anyway. I'd rather settle on one casing for .NET 6.

buyaa-n commented on src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:
```diff
@@ -29,7 +29,9 @@ public sealed partial class AssemblyBuilder : System.Reflection.Assembly
         public System.Reflection.Emit.ModuleBuilder? GetDynamicModule(string name) { throw null; }
         [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Types might be removed")]
         public override System.Type[] GetExportedTypes() { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFiles("This member throws an exception for assemblies embedded in a single-file app")]
```
```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFilesAttribute("This member throws an exception for assemblies embedded in a single-file app")]
```

buyaa-n commented on src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:
```diff
@@ -29,7 +29,9 @@ public sealed partial class AssemblyBuilder : System.Reflection.Assembly
         public override System.Type[] GetExportedTypes() { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFiles("This member throws an exception for assemblies embedded in a single-file app")]
         public override System.IO.FileStream GetFile(string name) { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFiles("This member throws an exception for assemblies embedded in a single-file app")]
```
```suggestion
        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFilesAttribute("This member throws an exception for assemblies embedded in a single-file app")]
```

buyaa-n commented on src/libraries/System.Reflection.Context/src/System/Reflection/Context/Delegation/DelegatingModule.cs:
```diff
@@ -22,6 +23,11 @@ public override Assembly Assembly
 
+        internal const string UnknownStringMessageInRAF = "Returns <unknown> for modules with no file path";
+
+#if NETCOREAPP
```
This might need `NET5_0_OR_GREATER`

buyaa-n commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Modules/RoModule.cs:
```diff
@@ -30,11 +31,21 @@ internal RoModule(string fullyQualifiedName)
         public sealed override Assembly Assembly => GetRoAssembly();
         internal abstract RoAssembly GetRoAssembly();
 
+
+
```
NIT extra line
```suggestion
```

buyaa-n commented on src/libraries/System.Reflection.MetadataLoadContext/src/System/Reflection/TypeLoading/Modules/RoModule.cs:
```diff
@@ -30,11 +31,21 @@ internal RoModule(string fullyQualifiedName)
+
+        internal const string UnknownStringMessageInRAF = "Returns <unknown> for modules with no file path";
+
+#if NETCOREAPP
```
Seems needs `#if NET5_0_OR_GREATER`
```suggestion
#if NET5_0_OR_GREATER
```

buyaa-n commented on src/coreclr/System.Private.CoreLib/src/System/Reflection/RuntimeAssembly.cs:
```diff
@@ -389,6 +390,8 @@ internal static RuntimeAssembly InternalLoad(AssemblyName assemblyName, ref Stac
                                   FileAccess.Read, FileShare.Read, FileStream.DefaultBufferSize, false);
         }
 
+
```
NIT
```suggestion
```

jkotas commented on src/coreclr/vm/commodule.cpp:
```diff
@@ -860,7 +860,7 @@ void QCALLTYPE COMModule::GetFullyQualifiedName(QCall::ModuleHandle pModule, QCa
                 retString.Set(fileName);
         } else {
-            retString.Set(W("<Unknown>"));
+            retString.Set(W("<unknown>"));
```
What is this trying to be consistent with? There is a lot of code out there that expects the current casing and that will break with this change. For example:

https://github.com/microsoft/perfview/blob/6b4771eecba5ed00555a34d70e19a8742af4ee0a/src/TraceEvent/NativeDlls.cs#L29

Also, the current casing is documented in the docs.

agocke replied:
Ugh, I had hoped this would be uncommon.

The case I found was in `RoModule.cs` as part of MetadataLoadContext. It uses lowercase `<unknown>`.

I'll revert the changes to `<Unknown>` for now, but I think MLC is inconsistent, so anyone relying on the behavior being consistent would be surprised. Maybe that's fine because the reflection APIs and MLC APIs don't cross over much.

jkotas replied:
I think it would be fine to fix MetadataLoadContext to use `<Unknown>`. I think it is unlikely that anybody depends on the exact MetadataLoadContext behavior in this case, so it should be fine to treat it as a simple MetadataLoadContext bugfix.

jkotas commented:
>  for change in casing in runtime. In theory this could be a breaking change for people upgrading from .NET 5 to .NET 6 but I think the inconsistent casing meant it was impossible for them to rely on specific behavior in this situation anyway.

I believe that the runtime reflection APIs have been consistently returning uppercase `<Unknown>` for images loaded as byte arrays since forever. I am not sure why you think it was impossible for apps and libraries to depend on this specific behavior.

agocke committed:
Revert "Found more missing annotations, clean up message and casing"

This reverts commit 283bd362f496016bc799a885cb4f7b193f49230c.

agocke committed:
Add annotations and address PR comments

agocke committed:
Apply suggestions from code review

Co-authored-by: Buyaa Namnan <buyankhishig.namnan@microsoft.com>

agocke committed:
Change casing of `<unknown>` for MetadataLoadContext

The current casing is `<unknown>` but the casing used in the runtime is
`<Unknown>`.

agocke committed:
Fix test baseline

agocke committed:
Merge remote-tracking branch 'upstream/main' into finish-annotating-raf

agocke committed:
Address new warnings

agocke committed:
Skip ScopeName for now and file a tracking bug

agocke commented:
OK, I think this is finally up-to-date. I ended up finding a questionable behavior in single-file for `Module.ScopeName` which needs to be investigated separately. I filed https://github.com/dotnet/runtime/issues/56519 to track that.

agocke commented:
@buyaa-n Anything else? I think the one tricky bit I've filed an issue for, so we can come back to it later.

buyaa-n commented on src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs:
```diff
@@ -29,7 +29,9 @@ public sealed partial class AssemblyBuilder : System.Reflection.Assembly
         public System.Reflection.Emit.ModuleBuilder? GetDynamicModule(string name) { throw null; }
         [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("Types might be removed")]
         public override System.Type[] GetExportedTypes() { throw null; }
+        [System.Diagnostics.CodeAnalysis.RequiresAssemblyFilesAttribute("This member throws an exception for assemblies embedded in a single-file app")]
```
NIT: wasn't it added to CodeBase too?
https://github.com/dotnet/runtime/blob/880b0d649d418d0453b3ebcf8dd44ecc19d1069e/src/libraries/System.Reflection.Emit/ref/System.Reflection.Emit.cs#L13

agocke replied:
I can add it there too -- this case is special-cased by the analyzer, so it's not necessary, just good style. I'll do that in a follow up PR.

buyaa-n replied:
Sounds good

buyaa-n reviewed (approved):
One NIT comment, overall LGTM
